src/py/flwr/server/superlink/state/in_memory_state.py

```diff
@@ -17,18 +17,35 @@
 
 import threading
 import time
+from dataclasses import dataclass
 from logging import ERROR
 from typing import Optional
 from uuid import UUID, uuid4
 
 from flwr.common import log, now
-from flwr.common.constant import NODE_ID_NUM_BYTES, RUN_ID_NUM_BYTES
-from flwr.common.typing import Run, UserConfig
+from flwr.common.constant import NODE_ID_NUM_BYTES, RUN_ID_NUM_BYTES, Status
+from flwr.common.typing import Run, RunStatus, UserConfig
 from flwr.proto.task_pb2 import TaskIns, TaskRes  # pylint: disable=E0611
 from flwr.server.superlink.state.state import State
 from flwr.server.utils import validate_task_ins_or_res
 
-from .utils import generate_rand_int_from_bytes, make_node_unavailable_taskres
+from .utils import (
+    generate_rand_int_from_bytes,
+    has_valid_sub_status,
+    is_valid_transition,
+    make_node_unavailable_taskres,
+)
+
+
+@dataclass
+class RunRecord:
+    """The record of a specific run, including its status and timestamps."""
+
+    run: Run
+    status: RunStatus
+    starting_at: str
+    running_at: str
+    finished_at: str
```

Comment:
Are `starting_at` and `finished_at` are the times? shouldn't we use float for them?
What is `running_at`?

Reply:
Yes, they are timestamps. The `running_at` is the time when the ServerApp starts.